### ISSUE TITLE
StoreBasedCartContext: Wrong Address Type if Adress Identifier is available

### ISSUE THREAD

**Issue by solverat** [coreshop/CoreShop#1627]
| Q                | A
| ---------------- | -----
| Bug report?      | yes
| Feature request? | no
| BC Break report? | no
| RFC?             | no

https://github.com/coreshop/CoreShop/blob/8a896977008af1b51ee856bfb678c1a8f0b517e7/src/CoreShop/Component/Core/Context/StoreBasedCartContext.php#L95-L99

Here the `StoreBasedCartContext` class assumes, that the default address always fits as shipping / invoice address, which is not always true. In some setups we need to follow EU laws (invoice => allow all countries, shipping => allow specific countries) and therefor we don't have "simple" addresses.

In this case, the default address identifier could be either `shipping` or `invoice` but never `null`. This needs to be checked first!

**Comment by solverat** [comment]
Same goes or `CartBlamerListener`: 

https://github.com/coreshop/CoreShop/blob/8a896977008af1b51ee856bfb678c1a8f0b517e7/src/CoreShop/Bundle/CoreBundle/EventListener/CartBlamerListener.php#L96-L102

**Comment by dpfaffenbauer** [comment]
any idea how to solve this?

**Comment by solverat** [comment]
Overwrites in my project:

```php
# CartBlameListener.php

private function blame(CustomerInterface $user)
{
    /**
     * @todo remove this class after #1627 has been fixed
     * @see  https://github.com/coreshop/CoreShop/issues/1627
     */

    $cart = $this->getCart();

    if ($cart === null) {
        return;
    }

    $cart->setCustomer($user);

    $defaultAddress = $user->getDefaultAddress();
    $shippingAddress = $user->getDefaultAddress();
    $invoiceAddress = $user->getDefaultAddress();

    if ($defaultAddress instanceof AddressInterface) {
        if ($defaultAddress->getAddressIdentifier() instanceof AddressIdentifierInterface) {
            if ($defaultAddress->getAddressIdentifier()->getName() === 'invoice') {
                $shippingAddress = null;
                $invoiceAddress = $user->getDefaultAddress();
            } elseif ($defaultAddress->getAddressIdentifier()->getName() === 'shipping') {
                $shippingAddress = $user->getDefaultAddress();
                $invoiceAddress = null;
            }
        }
    }

    if ($cart->getShippingAddress() === null) {
        $cart->setShippingAddress($shippingAddress);
    }

    if ($cart->getInvoiceAddress() === null) {
        $cart->setInvoiceAddress($invoiceAddress);
    }

    $this->cartProcessor->process($cart);

    if ($cart->getId()) {
        $this->cartManager->persistCart($cart);

        return;
    }

    $this->cartProcessor->process($cart);
}

```

```php
# StoreBasedCartContext.php

public function getCart()
{
    /**
     * @todo remove this class after #1627 has been fixed
     * @see https://github.com/coreshop/CoreShop/issues/1627
     */

    $cart = $this->cartContext->getCart();

    if (!$cart instanceof CartInterface) {
        throw new CartNotFoundException();
    }

    if ($cart->getInvoiceAddress() instanceof AddressInterface) {
        if ($cart->getInvoiceAddress()->getAddressIdentifier() instanceof AddressIdentifierInterface) {
            if ($cart->getInvoiceAddress()->getAddressIdentifier()->getName() !== 'invoice') {
                $cart->setInvoiceAddress(null);
            }
        }
    }

    if ($cart->getShippingAddress() instanceof AddressInterface) {
        if ($cart->getShippingAddress()->getAddressIdentifier() instanceof AddressIdentifierInterface) {
            if ($cart->getShippingAddress()->getAddressIdentifier()->getName() !== 'shipping') {
                $cart->setShippingAddress(null);
            }
        }
    }

    return $cart;
}
```

**Comment by dpfaffenbauer** [comment]
I think we should separate the address allocation into a separate service and call that from the Blamer and the Context, WDYT?

**Comment by benwalch** [comment]
that would be a indeed a better way, especially if one wants to override this 'default' behaviour of initial cart address assignments.

Btw, the overrides of the cart context turn out to be more difficult than it could be, bc of `final` class declarations
https://github.com/coreshop/CoreShop/blob/8a896977008af1b51ee856bfb678c1a8f0b517e7/src/CoreShop/Component/Core/Context/StoreBasedCartContext.php#L24
https://github.com/coreshop/CoreShop/blob/8a896977008af1b51ee856bfb678c1a8f0b517e7/src/CoreShop/Bundle/CoreBundle/EventListener/CartBlamerListener.php#L24
 and `private` methods
 https://github.com/coreshop/CoreShop/blob/8a896977008af1b51ee856bfb678c1a8f0b517e7/src/CoreShop/Component/Core/Context/StoreBasedCartContext.php#L91
 
Maybe I should address this in another issue.
